### PR TITLE
Support for *flat mapped* coherent secure RAM

### PR DESCRIPTION
**Pull request by etienne-lms**
This change enables secure coherent memory support and allow platform vexpress-qemu_virt to easily define a coherent memory at top of the secure RAM.

### CODE REVIEW

File: core/arch/arm/kernel/kern.ld.S [hunk 1]
```diff
@@ -53,9 +53,27 @@
 OUTPUT_FORMAT(CFG_KERN_LINKER_FORMAT)
 OUTPUT_ARCH(CFG_KERN_LINKER_ARCH)
 
+/*
+ * Start coherent RAM with a platform specific structure.
+ * Generic data to be place in coherent RAM can use attribute "coherent_ram".
+ */
+#define DECLARE_COHERENT_RAM_SECTION						\
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/kern.ld.S]: Why do we need a macro for this?

File: core/arch/arm/kernel/kern.ld.S [hunk 2]
```diff
@@ -356,3 +378,8 @@ SECTIONS
 	}
 
 }
+
+#if defined(CFG_TEE_COHERENT_START)
+ASSERT((__coherent_start & (4096 - 1)) == 0, "Coherent RAM start is not 4Kbyte aligned")
+ASSERT((__coherent_end & (4096 - 1)) == 0, "Coherent RAM end is not 4Kbyte aligned")
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/kern.ld.S]: Testing that `. = ALIGN(4096);` works?

File: core/arch/arm/plat-vexpress/platform_config.h [hunk 3]
```diff
@@ -222,6 +222,18 @@
 #define GICD_OFFSET		0
 #define GICC_OFFSET		0x10000
 
+/*
+ * CFG_TEE_COHERENT_SIZE can be used to define a coherent RAM at start of
+ * the secure RAM.
+ */
+#ifndef CFG_TEE_COHERENT_SIZE
+#define CFG_TEE_COHERENT_SIZE	0
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-vexpress/platform_config.h]: Why is this needed?

File: lib/libutils/ext/include/compiler.h [hunk 4]
```diff
@@ -54,6 +54,7 @@
 #define __rodata_unpaged __section(".rodata.__unpaged")
 #define __early_bss	__section(".early_bss")
 #define __noprof	__attribute__((no_instrument_function))
+#define __coherent	__attribute__((coherent_ram))
```
Comment (jenswi-linaro) [inline lib/libutils/ext/include/compiler.h]: My google karma is failing me, do you have a link to that attribute?

File: core/arch/arm/kernel/kern.ld.S [hunk 5]
```diff
@@ -53,9 +53,27 @@
...
 
+/*
+ * Start coherent RAM with a platform specific structure.
+ * Generic data to be place in coherent RAM can use attribute "coherent_ram".
+ */
+#define DECLARE_COHERENT_RAM_SECTION						\
+		. = CFG_TEE_COHERENT_START;					\
+		.coherent (NOLOAD) : {						\
+			__coherent_start = .;					\
+			KEEP(*(coherent_ram));					\
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/kern.ld.S]: All the other sections starts with a "." can't we have the same here?

File: core/arch/arm/kernel/kern.ld.S (hunk 1 above)
Comment (etienne-lms) [inline core/arch/arm/kernel/kern.ld.S]: To add some flexibility regarding the location of the coherent RAM, in case one wants it at a dedicated place. I proposed to possible location: before the load address or after the core reserved vaddr range (virt, no phys only, to fit with pager constraints). Since there were those 2 places, I preferred to use a single macro. You feel it bring complexity ?

File: core/arch/arm/kernel/kern.ld.S (hunk 5 above)
Comment (etienne-lms) [inline core/arch/arm/kernel/kern.ld.S]: my fault. i'll fix (see also comments on attribute `__coherent_ram`).

File: core/arch/arm/kernel/kern.ld.S [hunk 6]
```diff
@@ -53,9 +53,27 @@
 OUTPUT_FORMAT(CFG_KERN_LINKER_FORMAT)
 OUTPUT_ARCH(CFG_KERN_LINKER_ARCH)
 
+/*
+ * Start coherent RAM with a platform specific structure.
```
Comment (etienne-lms) [inline core/arch/arm/kernel/kern.ld.S]: Oups, this comment is related to the libpsci integ. I will remove it.

File: core/arch/arm/kernel/kern.ld.S (hunk 2 above)
Comment (etienne-lms) [inline core/arch/arm/kernel/kern.ld.S]: This is extra checks.  Actually `__coherent_end` is already aligned, and static mapping will assert `CFG_TEE_COHERENT_START` (hence `__coherent_start`) is already 4kB page aligned.
Those checks aimed at detecting issues at build time.

File: core/arch/arm/plat-vexpress/platform_config.h (hunk 3 above)
Comment (etienne-lms) [inline core/arch/arm/plat-vexpress/platform_config.h]: Default: no coherent RAM => null sized.

File: lib/libutils/ext/include/compiler.h (hunk 4 above)
Comment (etienne-lms) [inline lib/libutils/ext/include/compiler.h]: hmmm, will fix to `__section(".coherent_ram")`. Thanks.

File: core/arch/arm/plat-vexpress/platform_config.h (hunk 3 above)
Comment (jenswi-linaro) [inline core/arch/arm/plat-vexpress/platform_config.h]: May I suggest `CFG_TEE_COHERENT_SIZE ?= 0` in `core/arch/arm/arm.mk` instead?

File: core/arch/arm/kernel/kern.ld.S (hunk 1 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/kern.ld.S]: Aha, I didn't notice that. This is good then.
The `\` at the end of each line seems to have ended up at column 81 which obviously is beyond 80.

Comment (etienne-lms) [inline core/arch/arm/kernel/kern.ld.S]: Yep, i've noticed that from travis feedback. Bad luck.

File: core/arch/arm/kernel/kern.ld.S (hunk 2 above)
Comment (etienne-lms) [inline core/arch/arm/kernel/kern.ld.S]: Note: i will remove those extra checks.

Comment (jenswi-linaro) [inline core/arch/arm/kernel/kern.ld.S]: The assert that `. = ALIGN(4096);` works is just silly and and other assert could be done inside the `DECLARE_COHERENT_RAM_SECTION` macro instead.

File: core/arch/arm/plat-vexpress/platform_config.h (hunk 3 above)
Comment (etienne-lms) [inline core/arch/arm/plat-vexpress/platform_config.h]: Good idea. Thanks.

File: core/arch/arm/kernel/kern.ld.S (hunk 2 above)
Comment (etienne-lms) [inline core/arch/arm/kernel/kern.ld.S]: Adding several ASSERT() inside the macro `DECLARE_COHERENT_RAM_SECTION` combine with the 80 chars max line size make the macro not very easy to read. I will move all assertion at the end of the file. Nevertheless, fell free to comment back...

File: core/arch/arm/kernel/kern.ld.S [hunk 7]
```diff
@@ -379,7 +374,11 @@ SECTIONS
 
 }
 
-#if defined(CFG_TEE_COHERENT_START)
-ASSERT((__coherent_start & (4096 - 1)) == 0, "Coherent RAM start is not 4Kbyte aligned")
-ASSERT((__coherent_end & (4096 - 1)) == 0, "Coherent RAM end is not 4Kbyte aligned")
+#if CFG_TEE_COHERENT_SIZE
+ASSERT(!(__coherent_start & (4096 - 1)),
+					"Coherent memory start alignment");
+ASSERT(!(CFG_TEE_COHERENT_SIZE & (4096 - 1)),
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/kern.ld.S]: This and the following assert are useless, they are just checking that ` . = ALIGN(4096)` works.

File: core/arch/arm/kernel/kern.ld.S [hunk 8]
```diff
@@ -379,7 +374,11 @@ SECTIONS
 
 }
 
-#if defined(CFG_TEE_COHERENT_START)
-ASSERT((__coherent_start & (4096 - 1)) == 0, "Coherent RAM start is not 4Kbyte aligned")
-ASSERT((__coherent_end & (4096 - 1)) == 0, "Coherent RAM end is not 4Kbyte aligned")
+#if CFG_TEE_COHERENT_SIZE
+ASSERT(!(__coherent_start & (4096 - 1)),
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/kern.ld.S]: I don't see the problem with wrapping this inside the `DECLARE_COHERENT_RAM_SECTION` macros as:
```
                ASSERT(!(__coherent_start & (4096 - 1)),    \
                       "Coherent memory start alignment");     \
```

Comment (etienne-lms) [inline core/arch/arm/kernel/kern.ld.S]: An assertion in the macro could be ok, but the 3 make the macro a bit ugly (in my opinion).
But ok, I will move them (at least those really required) inside the macro.

File: core/arch/arm/kernel/kern.ld.S (hunk 7 above)
Comment (etienne-lms) [inline core/arch/arm/kernel/kern.ld.S]: The test above explicitly notifies developer that the coherent size follows the mapping constraint. Otherwise, this will be detected only a runtime. I found this explicit trace more convenient, but I agree it is redundant with the runtime assertions/panics.

Test below does not check the alignment but the size. In case one defines CFG_TEE_COHERENT_SIZE to a value below the effective coherent section computed at link stage, the coherent might the only partially mapped and only run time data abort could detect it.

File: core/arch/arm/plat-vexpress/platform_config.h [hunk 9]
```diff
@@ -222,7 +222,6 @@
 #define GICD_OFFSET		0
 #define GICC_OFFSET		0x10000
 
-
```
Comment (etienne-lms) [inline core/arch/arm/plat-vexpress/platform_config.h]: minor: useless change, to be discarded.

**Comment by etienne-lms** [comment]
Since the overall changes are quite small, I will rebase and squash the proposed commit to get a (hopefully) nice travis status and ease later reviews.

**Comment by etienne-lms** [comment]
thanks travis! i missed generic cases: must fix platform not defining any `CFG_TEE_COHERENT_START`.
I also found I forgot to push a change and let few crappy stuff on default location of the coherent memory. Fixes on-going...

**Comment by etienne-lms** [comment]
I will put this change on hold. 
My initial implementation was based on https://github.com/OP-TEE/optee_os/pull/1459 and I see now that it would imply too many conflicting changes to try to push it aside from this reference PR. Hence, I'll wait https://github.com/OP-TEE/optee_os/pull/1459 to mature and be merged before pushing back support for coherent memory.

**Comment by etienne-lms** [comment]
Rebased and adapted (on top of recently merged https://github.com/OP-TEE/optee_os/pull/1459).
Updated content: coherent memory is now always located before the core load address.

**Comment by etienne-lms** [comment]
rebased.

**Comment by jforissier** [comment]
Looks like this needs some rework?

**Comment by etienne-lms** [comment]
Yes, this deserves a rebase. Latest optee_os already defines coherent memory to maps rwx ram.
This p-r aimed at defining coherent mem (at least rw) and providing generic linking and mapping of the  coherent memory. Current p-r location coherent ram inside the tee_ram fla map constriant, for simplicity. Obviously https://github.com/OP-TEE/optee_os/pull/1729 locates coherent ram outside.

What's needs to be rebased:
- linking: declare a single coherent ram in **kern.lds.S** which can be above or below the flat mapped tee_ram) and define a `.coherent_ram` section attribute to allow linking code with the section.
- mapping: `register_phys_mem(COHERENT_MEM, ...);` should get the coherent mem to be mapped. This implies changes in [init_mem_map()#L845](https://github.com/OP-TEE/optee_os/blob/59ac3927b3bd14e133f57e7aa59278b09c2ca95d/core/arch/arm/mm/core_mmu.c#L845) to build a core vmem layout around coherent mem, `CFG_TEE_RAM_VA_SIZE` mapped and userland constraints.

**Comment by etienne-lms** [comment]
Rebased. 
- `init_mem_map()` sequence that defines the virtual memory layout is modified.
- `register_phys_mem(TEE_MEM_COHERENT)` maps the coherent.
- Defining `CFG_TEE_COHERENT_START/_SIZE` gets a `__coherent` section linked.
  (used by libpsci)

File: core/arch/arm/mm/core_mmu.c [hunk 10]
```diff
@@ -842,49 +900,118 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
...
+	 * other page tables.
+	 *
+	 * Note that non-LPAE cannot reuse secure xlat table for non secure
+	 * mapping entries. The non-LPAE will assign non secure entries with
+	 * small page region size after the vaspace.
+	 *
+	 * To speed up scans of the memory_map table, it is sorted, moving
+	 * all small-page mapped area before the pgdir mapped areas.
+	 */
+	qsort(memory_map, last, sizeof(struct tee_mmap_region),
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/mm/core_mmu.c]: I see your point. I will remove this sorting.

File: core/arch/arm/mm/core_mmu.c [hunk 11]
```diff
@@ -842,49 +900,118 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
...
+					  CORE_MMU_PGDIR_SIZE);
+	vaspace_size = ROUNDUP(vstart, CORE_MMU_PGDIR_SIZE) - vaspace_start;
+
+	/* All small page map area shall be assigned a location */
+	for (map = memory_map; !core_mmap_is_end_of_table(map); map++)
+		assert(map->va || map->region_size != SMALL_PAGE_SIZE);
+
+#ifdef CFG_WITH_LPAE
+	/* LPAE: 1 pgdir for protection, user memory located above the core */
+	va = CORE_MMU_PGDIR_SIZE;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/mm/core_mmu.c]: > So we start to map the rest at 2 MiB regardless of where flat mapped range is (it could even conflict with this)? 

The loop right after this init skips the virtual range already used to assigned virtual locations. This virtual area is expected to be fully cover the current value of `vaspace_start`/`vaspace_size`.

> More efficient use of level 2 tables would be to keep in same level 2 table as used above, it's 1 GiB so there should be plenty of space.

Agree. I will use the full 1GB around the flat map area to reuse the level2 xlat table.

File: core/arch/arm/mm/core_mmu.c [hunk 12]
```diff
@@ -90,6 +92,11 @@ static struct memaccess_area nsec_shared[] = {
 register_sdp_mem(CFG_TEE_SDP_MEM_BASE, CFG_TEE_SDP_MEM_SIZE);
 #endif
 
+#ifdef CFG_TEE_COHERENT_START
+register_phys_mem(MEM_AREA_TEE_COHERENT, CFG_TEE_COHERENT_START,
+					 CFG_TEE_COHERENT_SIZE);
+#endif
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c]: This should not be in commit "core: map registered coherent memory".
It should be in "core: linkable coherent memory section from CFG_TEE_COHERENT_START" which introduces `CFG_TEE_COHERENT_START`/`_SIZE`.

File: core/arch/arm/mm/core_mmu.c [hunk 13]
```diff
@@ -842,49 +871,115 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
...
+	/* Resuse xlat entries before vaspace */
+	init_smallpage_map(memory_map, vstart, vaspace_start, false);
+	vstart = vaspace_start + vaspace_size;
+
+	if (pgdir_sharing && coherent_start > vaspace_start) {
+		/* Resuse xlat entries between vaspace and coherent */
+		init_smallpage_map(memory_map, vstart, coherent_start, false);
+		vstart = coherent_start + coherent_size;
+	}
+
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c]: Looks like the one and the one below could be merged...

**Comment by etienne-lms** [comment]
Rebased (conflict to fix).
Tests in progress. qemu_virt ok, qemu_armv8 in progress... (in case ARMv8 would need coherent map support)

**Comment by jenswi-linaro** [comment]
Please squash the commits as you'd like to have them merged and I'll take a final look.

**Comment by etienne-lms** [comment]
`CFG_TEE_COHERENT_START`/`_SIZE` may not be a good label candidate.
- Any platform can get a coherent memory nicely mapped, using `register_phys_mem(MEM_AREA_TEE_COHERENT, ..., ...)`.
- Defining `CFG_TEE_COHERENT_START`/`_SIZE` makes the area a linked section inside the core implementation (allows static allocations inside coherent memory). 

Thus we may rather call this `CFG_LINKED_COHERENT_START`/`_SIZE`.

**Comment by jenswi-linaro** [comment]
It think the define is good as it is.
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
@MrVan, this p-r proposes a generic mapping support for coherent RAM.
From abut to be merge https://github.com/OP-TEE/optee_os/pull/1729, your coherent memory does not need flat mapping support. My change may conflict with your use of coherent memory.

**Comment by MrVan** [comment]
@etienne-lms I did not follow this p-r. I'll give a check and test with #1729.

**Comment by etienne-lms** [comment]
Thanks

**Comment by MrVan** [comment]
@etienne-lms 
It failed to boot correctly on i.mx7d-sdb
```
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: CFG_SHMEM_START type NSEC_SHM 0xbfe00000 size 0x00200000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: CFG_TA_RAM_START type TA_RAM 0xbe100000 size 0x01d00000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: VCORE_UNPG_RW_PA type TEE_RAM_RW 0xbe03d000 size 0x000c3000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: VCORE_UNPG_RX_PA type TEE_RAM_RX 0xbe000000 size 0x0003d000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: ROUNDDOWN(IRAM_S_BASE, CORE_MMU_DEVICE_SIZE) type TEE_COHERENT 0x00100000 size 0x00100000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: ROUNDDOWN(IRAM_BASE, CORE_MMU_DEVICE_SIZE) type TEE_COHERENT 0x00900000 size 0x00100000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: AIPS3_BASE type IO_SEC 0x30800000 size 0x00400000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: AIPS2_BASE type IO_SEC 0x30400000 size 0x00400000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: AIPS1_BASE type IO_SEC 0x30000000 size 0x00400000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: ANATOP_BASE type IO_SEC 0x30300000 size 0x00200000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:545: Physical mem map overlaps 0x30300000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: GIC_BASE type IO_SEC 0x31000000 size 0x00100000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: CONSOLE_UART_BASE type IO_NSEC 0x30800000 size 0x00200000
DEBUG:   [0x0] TEE-CORE:verify_special_mem_areas:470: No NSEC DDR memory area defined
DEBUG:   [0x0] TEE-CORE:add_va_space:571: type RES_VASPACE size 0x00a00000
DEBUG:   [0x0] TEE-CORE:add_va_space:571: type SHM_VASPACE size 0x02000000
ERROR:   [0x0] TEE-CORE: assertion '!coherent_start' failed at core/arch/arm/mm/core_mmu.c:858 <init_mem_map>
ERROR:   [0x0] TEE-CORE: Panic at core/kernel/assert.c:50 <_assert_break>
ERROR:   [0x0] TEE-CORE: Call stack:
ERROR:   [0x0] TEE-CORE:  0xbe004ec5
ERROR:   [0x0] TEE-CORE:  0xbe00a907
ERROR:   [0x0] TEE-CORE:  0xbe009eab
ERROR:   [0x0] TEE-CORE:  0xbe0055d1
ERROR:   [0x0] TEE-CORE:  0xbe0000c4
```

**Comment by jenswi-linaro** [comment]
+1 for `FLATMAP_COHERENT`

**Comment by etienne-lms** [comment]
Closing deprecate p-r.
